crates/red_knot_python_semantic/src/types.rs

```diff
@@ -71,55 +73,75 @@ fn symbol_by_id<'db>(db: &'db dyn Db, scope: ScopeId<'db>, symbol: ScopedSymbolI
             Some(
                 bindings_ty(db, use_def.public_bindings(symbol))
-                    .map(|bindings_ty| Symbol::Type(bindings_ty, use_def.public_boundness(symbol)))
+                    .map(|bindings_ty| {
+                        if let Some(boundness) = use_def.public_boundness(db, symbol) {
+                            Symbol::Type(bindings_ty, boundness)
+                        } else {
+                            Symbol::Unbound
+                        }
+                    })
                     .unwrap_or(Symbol::Unbound),
             )
-        } else {
-            None
-        };
-        // Intentionally ignore conflicting declared types; that's not our problem, it's the
-        // problem of the module we are importing from.
-
-        // TODO: Our handling of boundness currently only depends on bindings, and ignores
-        // declarations. This is inconsistent, since we only look at bindings if the symbol
-        // may be undeclared. Consider the following example:
-        // ```py
-        // x: int
-        //
-        // if flag:
-        //     y: int
-        // else
-        //     y = 3
-        // ```
-        // If we import from this module, we will currently report `x` as a definitely-bound
-        // symbol (even though it has no bindings at all!) but report `y` as possibly-unbound
-        // (even though every path has either a binding or a declaration for it.)
-
-        match undeclared_ty {
-            Some(Symbol::Type(ty, boundness)) => Symbol::Type(
-                declarations_ty(db, declarations, Some(ty)).unwrap_or_else(|(ty, _)| ty),
-                boundness,
-            ),
-            None | Some(Symbol::Unbound) => Symbol::Type(
-                declarations_ty(db, declarations, None).unwrap_or_else(|(ty, _)| ty),
-                Boundness::Bound,
-            ),
         }
-    } else {
-        bindings_ty(db, use_def.public_bindings(symbol))
-            .map(|bindings_ty| Symbol::Type(bindings_ty, use_def.public_boundness(symbol)))
-            .unwrap_or(Symbol::Unbound)
+        Some(Boundness::Bound) => None,
+    };
+
+    // Intentionally ignore conflicting declared types; that's not our problem, it's the
+    // problem of the module we are importing from.
+    let declarations = use_def.public_declarations(symbol);
+    match undeclared_ty {
+        Some(Symbol::Type(ty, boundness)) => Symbol::Type(
+            declarations_ty(db, declarations, Some(ty)).unwrap_or_else(|(ty, _)| ty),
+            boundness,
+        ),
+        None | Some(Symbol::Unbound) => Symbol::Type(
+            declarations_ty(db, declarations, None).unwrap_or_else(|(ty, _)| ty),
+            Boundness::Bound,
+        ),
     }
 }
 
 /// Shorthand for `symbol_by_id` that takes a symbol name instead of an ID.
 fn symbol<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> Symbol<'db> {
```

Comment:
This is not really related to your change but I'm surprised that `symbol` isn't a `salsa::tracked`... It is called across modules, but it might be fine, at least for as long as `symbol_by_id` never references any AST nodes (or anything that is directly derived from). 

However, it does seem like that `symbol_id` now calls `node_ref` methods in `StaticTruthiness`

https://github.com/astral-sh/ruff/blob/224e4b95fc0ee3d6a33a5f931dc57d180b949d86/crates/red_knot_python_semantic/src/types/static_truthiness.rs#L72-L76

We can't do that because it makes type inference of a single file dependent on the AST nodes of another file. 

This might already be a problem with how we use `Constraint`s today

Reply:
Okay, I think I was incorrect when I said that it isn't related to your changes. I did a quick review and it seems that the existing code all ends in a salsa query before accessing the constraint node. We should do the same for static thruthiness OR evaluate if we want to make `symbol_ty` a salsa query and remove the more fine-grained queries instead.